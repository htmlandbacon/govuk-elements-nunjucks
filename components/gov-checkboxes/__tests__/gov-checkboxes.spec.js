'use strict';

const nunjucks = require('nunjucks');
const cheerio = require('cheerio');

const checkBoxesTwo = [{id: 'one', label: 'One', value: 'one'}, {id: 'two', label: 'Two', value: 'two'}];

describe('gov-checkBoxes', () => {
  it('should set the ID as input-{name} if ID attr not supplied and generate checkboxes', () => {
    const name = 'selection';
    const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
      legend: 'Select a checkbox',
      checkBoxes: checkBoxesTwo,
      name
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`input-${name}-one`);
    expect(inputs.get(1).attribs.id).toBe(`input-${name}-two`);
    expect(output).toMatchSnapshot();
  });

  it('should use ID attribute value over generated input-{name} ID and generate check boxes', () => {
    const id = 'my-id';
    const name = 'selection';
    const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
      legend: 'Select a checkbox',
      checkBoxes: checkBoxesTwo,
      name,
      id
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`${id}-one`);
    expect(inputs.get(1).attribs.id).toBe(`${id}-two`);
    expect(output).toMatchSnapshot();
  });

  it('should add a form hint using the hint attribute', () => {
    const hint = 'select as many or as few as you like';
    const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
      legend: 'Select a checkbox',
      checkBoxes: checkBoxesTwo,
      name: 'selection',
      id: 'my-id',
      hint
    });
    const $ = cheerio.load(output);
    const hintText = $('legend span.form-hint').text();
    expect(hintText).toBe(hint);

    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`my-id-one`);
    expect(inputs.get(1).attribs.id).toBe(`my-id-two`);
    expect(output).toMatchSnapshot();
  });

  it('should set the check box to selected if value is set', () => {
    const checkBoxesTwoOneSelected = [{id: 'one', label: 'One', value: 'one'}, {id: 'two', label: 'Two', value: 'two', checked: 'true'}];
    const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
      legend: 'Select a checkbox',
      checkBoxes: checkBoxesTwoOneSelected,
      name: 'selection',
      id: 'my-id'
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    const firstCheckBox = inputs.get(0).attribs;
    const secondCheckBox = inputs.get(1).attribs;
    expect(firstCheckBox.id).toBe(`my-id-one`);
    expect(firstCheckBox.checked).toBe(undefined);
    expect(secondCheckBox.id).toBe(`my-id-two`);
    expect(secondCheckBox.checked).toBe('');
    expect(output).toMatchSnapshot();
  });

  it('should add error message and classes when passed an error object', () => {
    const error = 'Select an checkbox';
    const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
      legend: 'Select a checkbox',
      checkBoxes: checkBoxesTwo,
      name: 'selection',
      id: 'my-id',
      error
    });
    const $ = cheerio.load(output);

    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`my-id-one`);
    expect(inputs.get(1).attribs.id).toBe(`my-id-two`);

    const formGroupClasses = $('.form-group').attr('class');
    const errorMsg = $('legend span.error-message').text();

    expect(formGroupClasses).toBe('form-group form-group-error');
    expect(errorMsg).toBe(error);
    expect(output).toMatchSnapshot();
  });
  describe('checkbox options', () => {
    it('should add data fields when supplied', () => {
      const checkBoxWithData = [{id: 'one', label: 'One', value: 'one', dataTarget: 'data-target', dataJourneyClick: 'data-target-click'}];

      const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
        legend: 'Select a check box',
        checkBoxes: checkBoxWithData,
        id: 'my-id',
        name: 'selection'
      });

      const $ = cheerio.load(output);

      const firstDivContainer = $('div.multiple-choice');
      expect(firstDivContainer.get(0).attribs['data-target']).toBe('data-target');
      expect(firstDivContainer.get(0).attribs['data-journey-click']).toBe('data-target-click');

      const firstLabel = $('label');
      const oneCheck = expect.stringContaining('One');
      expect(firstLabel.text()).toEqual(oneCheck);
      const firstRadio = $('input').get(0).attribs;
      expect(firstRadio.id).toBe('my-id-one');
      expect(firstRadio.name).toBe('selection');
      expect(firstRadio.value).toBe('one');
      expect(output).toMatchSnapshot();
    });
  });
  describe('class options', () => {
    it('should add classes to legend span', () => {
      const classes = {legend: 'test-legend-class'};
      const output = nunjucks.render('./components/gov-checkboxes/template.njk', {
        legend: 'Select a check box',
        checkBoxes: checkBoxesTwo,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group legend span').attr('class');

      expect(labelClasses).toBe('form-label test-legend-class');
      expect(output).toMatchSnapshot();
    });
  });
});

