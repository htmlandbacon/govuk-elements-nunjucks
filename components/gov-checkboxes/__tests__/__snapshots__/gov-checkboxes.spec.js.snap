// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gov-checkBoxes checkbox options should add data fields when supplied 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a check box</span>
    </legend>
    <div class="multiple-choice" data-target="data-target" data-journey-click="data-target-click">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes class options should add classes to legend span 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label test-legend-class">Select a check box</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-two" type="checkbox" name="selection" value="two" />
      <label for="my-id-two">Two</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes should add a form hint using the hint attribute 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a checkbox</span>
      <span class="form-hint">select as many or as few as you like</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-two" type="checkbox" name="selection" value="two" />
      <label for="my-id-two">Two</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes should add error message and classes when passed an error object 1`] = `
<div class="form-group form-group-error" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a checkbox</span>
      <span class="error-message">Select an checkbox</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-two" type="checkbox" name="selection" value="two" />
      <label for="my-id-two">Two</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes should set the ID as input-{name} if ID attr not supplied and generate checkboxes 1`] = `
<div class="form-group" id="input-selection-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a checkbox</span>
    </legend>
    <div class="multiple-choice">
      <input id="input-selection-one" type="checkbox" name="selection" value="one"
      />
      <label for="input-selection-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="input-selection-two" type="checkbox" name="selection" value="two"
      />
      <label for="input-selection-two">Two</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes should set the check box to selected if value is set 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a checkbox</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-two" type="checkbox" name="selection" value="two" checked
      />
      <label for="my-id-two">Two</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-checkBoxes should use ID attribute value over generated input-{name} ID and generate check boxes 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label">Select a checkbox</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-one" type="checkbox" name="selection" value="one" />
      <label for="my-id-one">One</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-two" type="checkbox" name="selection" value="two" />
      <label for="my-id-two">Two</label>
    </div>
  </fieldset>
</div>
`;
